sympy/functions/elementary/piecewise.py

```diff
@@ -305,13 +306,87 @@ def doit(self, **hints):
+
+        Examples
+        ========
+        >>> from sympy import csc, log, Piecewise
+        >>> from sympy.abc import x
+        >>> f = Piecewise((csc(1/x), x < 1), (2*x**2 + 1, x < 5), (log(x), True))
+        >>> f.is_meromorphic(x, 0)
+        False
+        >>> f.is_meromorphic(x, 2)
+        True
+        >>> f.is_meromorphic(x, 10)
+        True
+
+        """
+        if not x.is_symbol:
+            raise TypeError("{} should be of symbol type".format(x))
+        a = sympify(a)
+        upper_bound_list, lower_bound_list = [], []
+        for l, u, e, i in self._intervals(x)[1]:
+            upper_bound_list.append(u)
+            lower_bound_list.append(l)
+
+        if a not in lower_bound_list and a not in upper_bound_list:
+            for bound in self.as_expr_set_pairs():
+                if a in bound[1]:
+                    return bound[0]._eval_is_meromorphic(x, a)
+        else:
+            return None
+
     def _eval_simplify(self, **kwargs):
         return piecewise_simplify(self, **kwargs)
 
     def _eval_as_leading_term(self, x, logx=None, cdir=0):
-        for e, c in self.args:
-            if c == True or c.subs(x, 0) == True:
-                return e.as_leading_term(x)
+        upper_bound_list, cond_list, lower_bound_list = [], [], []
+        for a, b, e, i in self._intervals(x)[1]:
+            upper_bound_list.append(b)
+            cond_list.append(e)
+            lower_bound_list.append(a)
+        upper = list(zip(cond_list, upper_bound_list))
+        lower = list(zip(cond_list, lower_bound_list))
+
+        # Check whether 0 is the upper boundary of a piece.
+        # If yes, we consider that piece and the next one for evaluation.
+        if S.Zero in upper_bound_list:
+            for i, pair in enumerate(upper):
+                if pair[1].is_zero:
+                    if cdir == -1:
+                        return pair[0].as_leading_term(x)
+                    else:
+                        return upper[i + 1][0].as_leading_term(x)
+
+        # Check whether 0 is the lower boundary of a piece.
+        # If yes, we consider that piece and the previous one for evaluation.
+        if S.Zero in lower_bound_list:
+            for i, pair in enumerate(lower):
+                if pair[1].is_zero:
+                    if cdir == -1:
+                        return lower[i - 1][0].as_leading_term(x)
+                    else:
+                        return pair[0].as_leading_term(x)
+
+        # Check whether 0 lies in the Interval of any piece.
+        # If yes, we consider that piece for evaluation.
+        for bound in self.as_expr_set_pairs():
+            if S.Zero in bound[1] and not S.Zero in bound[1].boundary:
+                return bound[0].as_leading_term(x)
+
+    def _at_infinity(self, x):
+        at_infinity = []
+        for i, pair in enumerate(self.as_expr_set_pairs()):
+            if pair[1].inf is S.NegativeInfinity or pair[1].sup is S.Infinity:
```

Comment:
Are these the first and last items of `self.as_expr_set_pairs()`? It so, then we could use the indices 0 and -1 without `enumerate`.

Reply:
Nope not necessarily first and last , it's basically taking out the pieces whose either lower boundary is `-oo` or upper boundary is `oo` . Check this out 
```
>>> p0 = Piecewise((0, Or(x < -2, x > 2)), (1, True))
>>> p0._at_infinity(x)
Piecewise((0, (x > 2) | (x < -2)))
```

Reply:
> not necessarily first and last

I would first take all boundary points in ascending order:

    -oo < a_1 < a_2 < ... < a_n < oo

Then the first and last intervals will be `(-oo, a_1)` and `(a_n, oo)`. These are not necessarily the same as the intervals defined by the first and last `ExprCondPair`s of a `Piecewise` expression. Some processing is needed for finding the boundary points.

Reply:
Yupp correct ! I realized that first we should always be covering the `(-oo, a_1)` interval and only then should we append the piece capturing `(a_n, oo)` . I'll make these changes asap !

Reply:
Now limit for p works as expected for all points, wasn't working properly at `oo` before 
```
>>> p = Piecewise((2, x > 2), (0, x < -2), (1, True))
>>> limit(p, x, 4)
2
>>> limit(p, x, oo)
2
>>> limit(p, x, -oo)
0
>>> limit(p, x, 2)
2
>>> limit(p, x, -2)
1
>>> limit(p, x, 2, '-')
1
>>> limit(p, x, -2, '-')
0
```